src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java

```diff
@@ -77,6 +91,37 @@ public void rebuildSessionFromFrame(MenuSession menuSession) throws CommCareSess
                         break;
                     }
                 }
+            } else if (screen instanceof QueryScreen) {
+                QueryScreen queryScreen = (QueryScreen)screen;
+                RemoteQueryDatum neededDatum = (RemoteQueryDatum) menuSession.getSessionWrapper().getNeededDatum();
+                boolean done = false;
+                for (StackFrameStep step : steps) {
+                    if (step.getId().equals(neededDatum.getDataId())) {
+                        URI uri = null;
+                        try {
+                            uri = new URI(step.getValue());
+                        } catch (URISyntaxException e) {
+                            e.printStackTrace();
+                            throw new CommCareSessionException("Query URL format error: " + e.getMessage());
+                        }
+                        try {
+                            ExternalDataInstance searchDataInstance = caseSearchHelper.getRemoteDataInstance(
+                                queryScreen.getQueryDatum().getDataId(),
+                                queryScreen.getQueryDatum().useCaseTemplate(),
+                                uri
+                            );
+                            queryScreen.setQueryDatum(searchDataInstance);
+                            screen = menuSession.getNextScreen(false);
+                            done = true;
+                        } catch (InvalidStructureException | IOException | XmlPullParserException | UnfullfilledRequirementsException e) {
+                            e.printStackTrace();
+                            throw new CommCareSessionException("Query response format error: " + e.getMessage());
+                        }
+                    }
+                }
+                if (done) {
```

Comment:
curious if this makes a difference since it's anyway at the end of the loop ?

Reply:
maybe this is meant to be inside the for loop above in which case it could be replaced with a `break`

Reply:
++ to Simon's suggestion

Reply:
Simon fixed in https://github.com/dimagi/formplayer/pull/1014/commits/275f5a095680ebeb632db30fc6ce3b8e6fbb72f2